### ISSUE TITLE
The trained model is transformed into pmml, but the prediction results are inconsistent

### ISSUE THREAD

**Issue by liuhuanshuo** [jpmml/sklearn2pmml#356]
Hello Villu,

I reread our previous communication and now I have successfully converted my trained model (lgb) to a `pmml` file。

While I was excited about this, a new problem appeared and I didn't know how to solve it or where it was.

### Problem description

To give you an idea of exactly what I'm having trouble with, let me describe my workflow again.

I have a trained lgb model that accepts 60 parameters and I have saved it as a.pkl file.

To convert it to a pmml file, I first read the model using joblib
```python
clf_1 = joblib.load(f'marketing_risk_lgb_1.pkl')
```

As mentioned above, this model accepts 60 features, but the original data has 61 columns (one feature is calculated by calculating two columns, yes, modify_days, which is calculated by modify_date and day_id).

So I performed the following feature conversion operation，In fact, only two columns actually operate; the rest are None
```
def make_modify_date_pipeline():
    return make_pipeline(ExpressionTransformer("X[0][:4] + '-' + X[0][4:6] + '-' + X[0][6:8] if len(X[0]) > 0 and X[0][0:8] < '20221230' else '2022-12-30'"), CastTransformer(dtype = "datetime64[D]"), DaysSinceYearTransformer(year = 2022))

def make_day_id_pipeline():
    return make_pipeline(ExpressionTransformer("X[1]"), CastTransformer(dtype = "datetime64[D]"), DaysSinceYearTransformer(year = 2022))

def make_feature_union():
    return FeatureUnion([
        ("modify_date", make_modify_date_pipeline()),
        ("day_id", make_day_id_pipeline())])


mapper_encode = [
    ('d3_daytime_opst_phone_called_cnt_rt',None),
    ('m1_accu_cnt',None),
    ('d1_under15s_opst_phone_cnt_rt',None),
    ('d15_creditcard_call_cnt_rt',None),
    ('d1_daytime_voice_cnt_rt',None),
    ('d15_under15s_called_opst_mbl_cnt',None),
    ('d3_dur_under_10s_cnt_rt',None),
    ('d1_accu_rm_cnt_rt',None),
    ('d3_accu_rm_called_dur',None),
    ('d1_once_called_opst_phone_cnt_rt',None),
    ('d3_once_called_opst_mbl_cnt',None),
    ('d1_once_called_opst_mbl_cnt',None),
    ('d3_called_dur_under_30s_cnt_rt',None),
    ('d15_nighttime_mbl_innet_cnt',None),
    ('d3_daytime_voice_called_cnt_rt',None),
    ('d3_called_dur_under_10s_cnt_rt',None),
    ('d3_dur_over_180s_cnt',None),
    ('m1_accu_called_cnt',None),
    ('d1_daytime_voice_called_cnt_rt',None),
    ('d7_loc_called_inact_day_cnt',None),
    ('d3_under15s_called_opst_mbl_cnt_rt',None),
    ('m3_accu_called_cnt',None),
    ('d1_called_dur_under_30s_cnt_rt',None),
    ('d15_rm_inact_day_cnt',None),
    ('d3_daytime_opst_mbl_called_cnt_rt',None),
    ('d3_called_opst_phone_cnt_rt',None),
    ('d3_under15s_called_opst_phone_cnt_rt',None),
    ('m3_accu_cnt',None),
    ('d7_under15s_called_opst_phone_cnt_rt',None),
    ('d3_under15s_opst_phone_cnt_rt',None),
    ('m3_accu_loc_cnt',None),
    ('d15_called_dur_over_180s_cnt',None),
    ('d3_accu_loc_cnt_rt',None),
    ('d1_once_opst_phone_cnt_rt',None),
    ('d3_daytime_opst_phone_cnt_rt',None),
    ('d1_dur_under_30s_cnt_rt',None),
    ('d7_under15s_called_opst_mbl_cnt_rt',None),
    ('d1_accu_loc_cnt_rt',None),
    ('d1_daytime_opst_phone_cnt_rt',None),
    ('d3_under15s_opst_mbl_cnt',None),
    ('d1_accu_called_cnt_rt',None),
    ('d15_bankloan_call_cnt_rt',None),
    ('d1_dur_under_10s_cnt_rt',None),
    ('d7_once_called_opst_mbl_cnt_rt',None),
    ('d7_dur_over_180s_cnt',None),
    (['modify_date','day_id'],[make_feature_union(), ExpressionTransformer("X[1] - X[0] if X[1]>X[0] else -1")],{'alias':'modify_days'}),
    ('d7_once_called_opst_mbl_cnt',None),
    ('d3_once_opst_phone_cnt_rt',None),
    ('d1_daytime_opst_phone_called_cnt_rt',None),
    ('d15_mbl_innet_day',None),
    ('d3_once_called_opst_phone_cnt_rt',None),
    ('d15_express_call_cnt_rt',None),
    ('d15_insurance_call_cnt_rt',None),
    ('d1_daytime_opst_mbl_called_cnt_rt',None),
    ('d1_accu_loc_called_dur',None),
    ('d1_called_opst_phone_cnt_rt',None),
    ('m3_accu_loc_called_cnt',None),
    ('d7_called_dur_over_180s_cnt',None),
    ('d3_nighttime_mbl_innet_cnt',None),
    ('channel_type_cd_3',LookupTransformer(mapping = all_cate_dict['channel_type_cd_3'], default_value = 0.0))]
```
Next build pipeline
```
mapper = DataFrameMapper(mapper_encode, input_df=True)
pipeline_test = PMMLPipeline(
    steps=[("mapper", mapper),
           ("classifier", clf_1)])

pipeline_test.target_fields = ["my_single_target"]
```

So far, everything is working properly and the pipeline I built can get the correct prediction results!
```python
pipeline_test.predict_proba(x_oot_1)

array([[0.19848053, 0.80151947],
       [0.16321889, 0.83678111],
       [0.09590413, 0.90409587],
       ...,
       [0.14002773, 0.85997227],
       [0.17470267, 0.82529733],
       [0.14859565, 0.85140435]])
```

I can also convert it to pmml successfully
```
sklearn2pmml(pipeline_test,'pipeline_test.pmml')
```

But if I use `pypmml` to read this file, and use the same data to predict, and get the wrong result, maybe not wrong, **the overall result is similar, but there is a certain gap with the right result,For example, 0.84 becomes 0.81**
```
from pypmml import Model
lgb_test = Model.load('pipeline_test.pmml')
lgb_test.predict(x_oot_1)

	probability(0)	probability(1)
0	0.152631	0.847369
1	0.139494	0.860506
2	0.067159	0.932841
3	0.122585	0.877415
4	0.068622	0.931378
...	...	...
20454	0.073462	0.926538
20455	0.070421	0.929579
20456	0.159350	0.840650
20457	0.106328	0.893672
20458	0.141541	0.858459
```

## Possible problems?

From the above process, we can see that there is no problem from the whole process to the pipeline, and the result predicted by only using the pipeline is in line with the expectation!

So I locate the problem to the pmml transformation!

I tried searching related posts, it seems that pmml ignores some features of less importance?

Of course, I am looking forward to your comments and reply on the specific reasons!

**Comment by liuhuanshuo** [comment]
I'm sorry that I can't give you the test data like other issues.

All my problems can be summed up in one sentence. If the pipeline works properly, just converting the pipeline to pmml, the same data prediction results will be inconsistent.

Also, if I save the pipeline as a pkl.z file and use joblib to read and predict it, I can get the desired result!

**Comment by vruusmann** [comment]
> But if I use pypmml to read this file, and use the same data to predict, and get the wrong result, maybe not wrong, the overall result is similar, but there is a certain gap with the right result,For example, 0.84 becomes 0.81

Do not use the PyPMML package! Use the [JPMML-Evaluator-Python](https://github.com/jpmml/jpmml-evaluator-python) package instead.

Does the error persist when executing `Evaluator.evaluateAll(X)` method (instead of `Model.predict(X)`)?

**Comment by vruusmann** [comment]
TLDR: This is not a conversion-side problem. This is a scoring-side problem.

**Comment by vruusmann** [comment]
Closing as invalid - I don't do support for third-party software.

Maybe PyPMML simply messes up input column mappings (for some reason it uses a positional mapping, not name-based mapping). Maybe it can't do datetime arithmetics right.

Not my problem. The JPMML software project provides full end-to-end toolset, which is demonstrably correct.

**Comment by liuhuanshuo** [comment]
> Do not use the PyPMML package! Use the [JPMML-Evaluator-Python](https://github.com/jpmml/jpmml-evaluator-python) package instead.

Thank you very much for your advice. I didn't even know there was a tool like JPMML-Evaluator-Python that could be used to read pmml files.

I would recommend that my team all switch to jpmml related tools.

But I am having some problems with JPMML-Evaluator-Python, and I will raise issues under that project! Anyway, sklearn2pmml has succeeded in getting me to convert pmml files, awesome!
